backend/onedrive/onedrive.go

```diff
@@ -1503,7 +1503,33 @@ func (f *Fs) PublicLink(ctx context.Context, remote string, expire fs.Duration,
 		fmt.Println(err)
 		return "", err
 	}
-	return result.Link.WebURL, nil
+
+	url := result.Link.WebURL
+
+	// Convert share link to direct download link if target is not a folder
+	chunks := strings.Split(url, "/")
+	if f.driveType == driveTypePersonal && chunks[3] != "u" {
```

Comment:
You cannot expect that url path will always have 3+ segments.
Sooner or later Microsoft will change something and rclone will panic.
Next, the term "chunk" in rclone is usually related to data bytes, let's use something else.
Also log something loud if all your conditions failed.
Please rewrite this snippet like this:
```
segments := strings.Split(url, "/")
linkType := ""
if len(segments) >= 3 {
  linkType = segments[3]
}
switch {
case f.driveType == driveA && linkType == urlB:
  url = ...
case ...:
  ...
default:
  fs.Logf(..., "Unsupported link type. Cannot convert to downloadable URL")
}
```

Reply:
Thanks for your review. Suggestions of terms and logs are great. But I'm a little bit confused about the panic handling. I have two concerns:
1. If Microsoft did change its format, this function will return something differently. Instead of waiting for people to notice that something is wrong, wouldn't it be nice to just let it panic so the problem can be noticed and fixed immediately?
2. If we do want to handle some changes, how about rewriting the snippet like this to handle it more properly:
```go
       	url := result.Link.WebURL
        segments := strings.Split(url, "/")
	switch {
	case f.driveType == driveTypePersonal:
		// https://stackoverflow.com/questions/37951114/direct-download-link-to-onedrive-file
		if len(segments) != 5 {
			fs.Logf(f, "Microsoft might change URL format. Cannot convert to downloadable URL")
		} else if segments[3] != "u" {
			enc := base64.StdEncoding.EncodeToString([]byte(url))
			url = "https://api.onedrive.com/v1.0/shares/u!" + enc[:len(enc)-1] + "/root/content"
		}
	case ...:
		...
	default:
		fs.Logf(...)
```
In the case of format changes, the original URL will be used.

Reply:
> 1. ... let it panic so the problem can be noticed...

panic == crash, i.e. wrong program
instead if crashing we must log error

> 2. If we do want to handle some format changes, how about rewriting the snippet like this to handle it more properly:
> Since some case will access up to the 7th segment.
> ```go
> url := result.Link.WebURL
> segments := strings.Split(url, "/")
> switch {
> case f.driveType == driveTypeBusiness:
>     if len(segments) != 8 {
>         fs.Logf(f, "Conversion method is outdated. Cannot convert to downloadable URL")
>     } else if segments[3] != ":f:" {
>         url = strings.Join(segments[0:3], "/") + "/" + segments[5] + "/" + segments[6] +
>             "/_layouts/15/download.aspx?share=" + segments[7]
>     }

before accessing segment #3 this fragment checks that no. of segments is 8, so OK

> case ...:
>     ...
> default:
>     fs.Logf(...)
> ```
> In the case of format changes, the original URL will be used.